src/xwayland.c

```diff
@@ -33,26 +33,67 @@ xwayland_view_get_size_hints(struct view *view)
+
+	switch (wlr_xwayland_icccm_input_model(xsurface)) {
+	/*
+	 * Abbreviated from ICCCM section 4.1.7 (Input Focus):
+	 *
+	 * Passive Input - The client expects keyboard input but never
+	 * explicitly sets the input focus.
+	 * Locally Active Input - The client expects keyboard input and
+	 * explicitly sets the input focus, but it only does so when one
+	 * of its windows already has the focus.
+	 *
+	 * Passive and Locally Active clients set the input field of
+	 * WM_HINTS to True, which indicates that they require window
+	 * manager assistance in acquiring the input focus.
+	 */
+	case WLR_ICCCM_INPUT_MODEL_PASSIVE:
+	case WLR_ICCCM_INPUT_MODEL_LOCAL:
 		return true;
-	}
+
 	/*
-	 * Paraphrased from ICCCM section 4.1.7 (Input Focus):
+	 * Globally Active Input - The client expects keyboard input and
+	 * explicitly sets the input focus, even when it is in windows
+	 * the client does not own. ... It wants to prevent the window
+	 * manager from setting the input focus to any of its windows
+	 * [because it may or may not want focus].
 	 *
-	 * Clients set the input field of WM_HINTS to True to indicate
-	 * that they require window manager assistance in acquiring the
-	 * input focus. Clients set the input field to False to request
-	 * that the window manager not set the input focus to their
-	 * top-level window.
+	 * Globally Active client windows may receive a WM_TAKE_FOCUS
+	 * message from the window manager. If they want the focus, they
+	 * should respond with a SetInputFocus request.
 	 *
-	 * Clients that use XSetInputFocus() to explicitly set the input
-	 * focus should set the WM_TAKE_FOCUS atom in WM_PROTOCOLS.
-	 * Currently, labwc does not support this method of taking focus
-	 * and thus ignores WM_TAKE_FOCUS. These views can still be
-	 * focused by explicit user action (e.g. clicking in them).
+	 * [Currently, labwc does not fully support clients voluntarily
+	 * taking focus via the WM_TAKE_FOCUS + SetInputFocus mechanism
+	 * and avoids automatically focusing Globally Active windows.
+	 * This may change in future.]
 	 */
-	return (bool)hints->input;
+	case WLR_ICCCM_INPUT_MODEL_GLOBAL:
+		/*
+		 * Assume the window does want focus if it wants window
+		 * decorations (according to _MOTIF_WM_HINTS). This is
+		 * a stop-gap fix to ensure that various applications
+		 * (mainly Java-based ones such as IntelliJ IDEA) get
+		 * focus normally and appear in the window switcher. It
+		 * would be better to match based on _NET_WM_WINDOW_TYPE
+		 * instead, but that property isn't currently available
+		 * through wlroots API.
+		 */
+		return (xsurface->decorations ==
+			WLR_XWAYLAND_SURFACE_DECORATIONS_ALL);
+
+	/*
+	 * No Input - The client never expects keyboard input.
+	 *
+	 * No Input and Globally Active clients set the input field to
+	 * False, which requests that the window manager not set the
+	 * input focus to their top-level window.
+	 */
+	default: /* WLR_ICCCM_INPUT_MODEL_NONE */
+		return false;
+	}
```

Comment:
Based on the comment: why is `WLR_ICCCM_INPUT_MODEL_NONE` not a `case` here?
Also, no need to check for the input hint here anymore?

Reply:
Just a style idiosyncrasy of mine, I guess. Some compilers and/or style guidelines complain if you have a `default` case but all enum values are covered. Others complain if you don't have a `default` case. This makes both happy. I don't care strongly about it.

The return value of `wlr_xwayland_icccm_input_model()` incorporates `WM_HINTS.input` already. See:
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/xwayland/xwm.c?ref_type=heads#L2318

Reply:
I think using the case and no default here is the safest variant. If wlroots ends up adding a new member to the enum we should run into a compile error as not all enums are covered and there is not default. That prevents accidentally treating a newly added enum as `WLR_ICCCM_INPUT_MODEL_NONE`.

> The return value of wlr_xwayland_icccm_input_model() incorporates WM_HINTS.input already. 

Got it, thanks for the explanation.